staging/src/k8s.io/apimachinery/pkg/util/net/interface.go

```diff
@@ -406,18 +424,19 @@ func chooseHostInterfaceFromRoute(routes []Route, nw networkInterfacer) (net.IP,
 	return nil, fmt.Errorf("unable to select an IP from default routes.")
 }
 
-// If bind-address is usable, return it directly
-// If bind-address is not usable (unset, 0.0.0.0, or loopback), we will use the host's default
-// interface.
+// ChooseBindAddress returns an IP address for a daemon, based on the given bindAddress:
+// If bindAddress is unset, it returns the host's default IP, as with ChooseHostInterface().
+// If bindAddress is unspecified or loopback, it returns the default IP of the same
+// address family as bindAddress.
```

Comment:
> it returns the default IP of the same address family as bindAddress.

does this match the stdlib bind behavior? if I bind to `0.0.0.0`, do I only get bound to all ipv4 interfaces?

also, don't we use 0.0.0.0 as a stand-in for "all" in some of our default options? what would this change do to defaults on an ipv6-only machine?

Reply:
> does this match the stdlib bind behavior? if I bind to 0.0.0.0, do I only get bound to all ipv4 interfaces?

The PR doesn't change what address the apiserver actually binds to. The function name is misleading. (The apiserver always binds to the address you passed in as `--bind-address`. `ChooseBindAddress()` is used to figure out the default value for `--advertise-address`.) Probably we should rename the function...

> also, don't we use "0.0.0.0" as a stand-in for "all" in some of our default options? what would this change do to defaults on an ipv6-only machine?

Ah, yes, that would break with this PR. I guess for compatibility it should *prefer* the passed-in address family, but not require it.

Reply:
> I guess for compatibility it should prefer the passed-in address family, but not require it.

isn't this already handled by the logic on line 420-421?